File: frontend/src/machineComposition.ts
import type { MachinePartType, RacingType } from "./types";

export const machineTypes: RacingType[] = ["SPEED", "ACCELERATION", "HANDLING", "POWER", "BOOST"];

export function requiredMachineSlots(type: RacingType | null): MachinePartType[] {
  if (!type) return [];
  return type === "BOOST" ? ["FRONT", "REAR"] : ["FRONT", "REAR", "TIRE"];
}

export function machineTypeLabel(type: RacingType | null): string {
  if (!type) return "Unknown";
  return type === "BOOST" ? "Boost · Extreme Gear" : type[0] + type.slice(1).toLowerCase();
}

File: backend/src/main/java/dev/ringlab/domain/gamedata/MachineComposition.java
package dev.ringlab.domain.gamedata;

import java.util.List;

/** RingLab's current CrossWorlds machine policy; never applied to racer types. */
public final class MachineComposition {
  private MachineComposition() {}

  public static List<MachinePartType> requiredSlots(RacingType machineType) {
    if (machineType == null) throw new IllegalArgumentException("Machine racing type is unknown");
    return machineType == RacingType.BOOST
        ? List.of(MachinePartType.FRONT, MachinePartType.REAR)
        : List.of(MachinePartType.FRONT, MachinePartType.REAR, MachinePartType.TIRE);
  }
}
